[View in Colaboratory](https://colab.research.google.com/github/mahedjaved/tutorial-rep/blob/master/1_GettingStarted_EagerExec.ipynb)

# Chapter 1 - Getting Start - Exec

* Currently many APIs in tensorflow
* EagerExec is the easiest high level API in TF

In [0]:
#!pip install -q --upgrade tensorflow
#!pip freez
#!pip install mxnet
#!pip uninstall tf-nightly 

### Basic Guid for Eager Execution

* Eager execution enables tensorflow to turn off **computational graph mode** and begin with imperative execution

---
Begin by importing libraries


In [2]:
from __future__ import absolute_import, division, print_function

import os
currDir=os.getcwd()
import matplotlib.pyplot as plt

import tensorflow as tf
import tensorflow.contrib.eager as tfe

tf.enable_eager_execution()

print("TensorFlow version: {}".format(tf.VERSION))
print("Eager execution: {}".format(tf.executing_eagerly()))

TensorFlow version: 1.8.0
Eager execution: True


### Iris Classifier

* A more modest approach to classifying flowers
* Mostly based on sizes
* Will only classify 3 species
    * Iris setosa
    * Iris virginica
    * Iris versicolor

#### 1. Download the dataset

In [5]:
train_dataset_url = "http://download.tensorflow.org/data/iris_training.csv"

train_dataset_fp = tf.keras.utils.get_file(fname=os.path.basename(train_dataset_url),
                                           origin=train_dataset_url)

print("Local copy of the dataset file: {}".format(train_dataset_fp))

8192/2194 [================================================================================================================] - 0s 0us/step
Local copy of the dataset file: /content/.keras/datasets/iris_training.csv


#### 2. Inspect the dataset

* The '-n5' command takes a peak at only the first five entries

In [6]:
!head -n5 {train_dataset_fp}

120,4,setosa,versicolor,virginica
6.4,2.8,5.6,2.2,2
5.0,2.3,3.3,1.0,1
4.9,2.5,4.5,1.7,2
4.9,3.1,1.5,0.1,0


* The first line contains header information e.g. 120 points to the size of the dataset 
* The following rows capture features
* E.g. the first four columns capture the measurements of the flower
* The last layer captures the category type defining the name of the flower

#### 3. Parsing dataset 

* Parsing allows dataset to be readable for python from CSV format
* Each row of the .csv file is passed into the function "parse_csv" as an argument
* Finally the last field is parsed as a label {field=column}
* The function "parse_csv" combines fields into a single tensor := each column is recorded as a single tensor
* And it returns both "features" and "labels"

In [0]:
def parse_csv(line):
  example_defaults=[[0.], [0.], [0.], [0.], [0.]]
  parsed_line=tf.decode_csv(line, example_defaults)
  # first 4 fields are features, combined into single tensor
  features=tf.reshape(parsed_line[:-1], shape=(4,))
  # set last field as label
  label=tf.reshape(parsed_line[:-1], shape=())
  return features, label

#### 4. Create training data using tf.data API